### PR TITLE
Set an upper limit on CPU threads in distributed training

### LINKED ISSUE #16737: Extreme single thread cpu kernel usage while training on GPU
### Bug description

As stated [in this Pytorch forum post](https://discuss.pytorch.org/t/extreme-single-thread-cpu-kernel-usage-while-training-on-gpu/172380).
As the title and images below suggest, during training, one of the cpu cores is experiencing extreme kernel usage while other cores barely moves. Despite I am using GPU as the accelerator.
image
image
I had tried to set the num_workers in the dataloader to 2 or more, make no difference. Set pin_memory to True or False doesn’t seem to do anything as well.
torch.get_num_threads() returns 6, but torch.set_num_threads() = 6 doesn’t help with the issue.
If I switch the accelerator from cuda to cpu, the model also only uses one cpu core for training, which is extremely inefficient.
The model itself is a Half-UNet for image segmentation, 16bit grey scale image data are converted into tensors in the Dataset, before a variety of transformations got applied and send into the model for training.

export OMP_NUM_THREADS=6 doesn’t seem to fix the issue. I also tried transferring all the data in the dataloader to the cuda device in the `__init__` function as well, it marginally improved the rate of training (if any), but doesn’t help with my issue.

Please see the original post for more information.

### How to reproduce the bug

```python
Please see the original post.
```

### Error messages and logs

_No response_

### Environment

<details>
  <summary>Current environment</summary>

```
* CUDA:
	- GPU:
		- NVIDIA GeForce GTX 1050 Ti
	- available:         True
	- version:           11.7
* Lightning:
	- lightning-utilities: 0.6.0.post0
	- pytorch-lightning: 1.9.0
	- torch:             1.13.1
	- torchaudio:        0.13.1
	- torchmetrics:      0.11.1
	- torchvision:       0.14.1
* Packages:
	- absl-py:           1.3.0
	- aiohttp:           3.8.3
	- aiosignal:         1.3.1
	- appdirs:           1.4.4
	- async-timeout:     4.0.2
	- attrs:             22.2.0
	- blinker:           1.4
	- bottleneck:        1.3.5
	- brotlipy:          0.7.0
	- cachetools:        4.2.2
	- certifi:           2022.12.7
	- cffi:              1.15.1
	- charset-normalizer: 2.0.4
	- click:             8.0.4
	- cryptography:      38.0.4
	- flit-core:         3.6.0
	- frozenlist:        1.3.3
	- fsspec:            2023.1.0
	- google-auth:       2.6.0
	- google-auth-oauthlib: 0.4.4
	- grpcio:            1.42.0
	- idna:              3.4
	- imageio:           2.19.3
	- lightning-utilities: 0.6.0.post0
	- markdown:          3.4.1
	- markupsafe:        2.1.1
	- mkl-fft:           1.3.1
	- mkl-random:        1.2.2
	- mkl-service:       2.4.0
	- multidict:         6.0.4
	- numexpr:           2.8.4
	- numpy:             1.23.5
	- oauthlib:          3.2.1
	- packaging:         23.0
	- pandas:            1.5.2
	- pillow:            9.3.0
	- pip:               22.3.1
	- pooch:             1.4.0
	- protobuf:          3.20.3
	- pyasn1:            0.4.8
	- pyasn1-modules:    0.2.8
	- pycparser:         2.21
	- pyjwt:             2.4.0
	- pyopenssl:         22.0.0
	- pysocks:           1.7.1
	- python-dateutil:   2.8.2
	- pytorch-lightning: 1.9.0
	- pytz:              2022.7
	- pyyaml:            6.0
	- requests:          2.28.1
	- requests-oauthlib: 1.3.0
	- rsa:               4.7.2
	- scipy:             1.10.0
	- setuptools:        65.6.3
	- six:               1.16.0
	- tensorboard:       2.10.0
	- tensorboard-data-server: 0.6.1
	- tensorboard-plugin-wit: 1.8.1
	- torch:             1.13.1
	- torchaudio:        0.13.1
	- torchmetrics:      0.11.1
	- torchvision:       0.14.1
	- tqdm:              4.64.1
	- typing-extensions: 4.4.0
	- urllib3:           1.26.14
	- werkzeug:          2.2.2
	- wheel:             0.37.1
	- yarl:              1.8.2
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- ELF
	- processor:         x86_64
	- python:            3.10.9
	- version:           #64-Ubuntu SMP Thu Jan 5 11:43:13 UTC 2023

```

</details>

### More info

_No response_

cc @carmocca @justusschock @awaelchli @borda

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #16737

Aligns our default launcher closer with torchrun's settings regarding the number of threads:
https://github.com/pytorch/pytorch/blob/e0be9ebc181fbf7b2ed4e641cd24a0dadd063f27/torch/distributed/run.py#L702-L714

## Benchmarks
All experiments on 8xA100 40GB machine, 256 CPU cores, PyTorch 2.2.0.dev20230920+cu121 nightly. 

### Computer Vision Example
```py
import os

import lightning.pytorch as pl
import timm
import torch
from torch.nn import functional as F
from torch.utils.data import DataLoader
from torchvision import transforms as tfs
from torchvision.datasets import CIFAR10
import time

epochs = 5
epoch_size = 100


class LitModel(pl.LightningModule):
    def __init__(self):
        super().__init__()
        self.model = timm.create_model("rexnet_150", pretrained=True, num_classes=10)
        self.t0 = 0
        self.epoch_time = []

    def configure_optimizers(self):
        return torch.optim.Adam(self.parameters(), lr=0.1)

    def on_train_start(self):
        # torch.set_num_threads(1)
        # torch.set_num_interop_threads(1)
        print(os.environ.get("OMP_NUM_THREADS"))
        print(torch.get_num_threads(), torch.get_num_interop_threads())

    def on_train_epoch_start(self):
        self.t0 = time.perf_counter()

    def on_train_epoch_end(self):
        self.epoch_time.append(time.perf_counter() - self.t0)

    def on_train_end(self):
        avg_epoch_time = sum(self.epoch_time) / len(self.epoch_time)
        it_per_sec = epochs * epoch_size / sum(self.epoch_time)
        self.trainer.print(f"avg time per epoch: {avg_epoch_time:.3f}")
        self.trainer.print(f"it/s: {it_per_sec:.3f}")

    def training_step(self, batch):
        x, y = batch
        logits = self.model(x)
        return F.cross_entropy(logits, y)


def run():
    transform = tfs.Compose([tfs.Resize((224, 224)), tfs.ToTensor()])
    dataset = CIFAR10(".", train=True, download=True, transform=transform)
    dataloader = DataLoader(dataset, batch_size=64)
    model = LitModel()
    trainer = pl.Trainer(max_epochs=epochs, limit_train_batches=epoch_size, accelerator="cuda", devices=2, strategy="ddp")
    trainer.fit(model, dataloader)


if __name__ == "__main__":
    run()
```

Single GPU

| num_threads | it/s |
| --- | --- |
| cpu_count | 0.30 |
| cpu_count / num_procs | 9.673 |
| 1 | 9.634 |

DDP with 2 processes/GPUs

| num_threads | it/s |
| --- | --- |
| cpu_count | 0.16 |
| cpu_count / num_procs | 5.940 |
| 1 | 7.640 |

### Transformer Example

```py
import lightning as L
import torch
import torch.nn.functional as F
from lightning.pytorch.demos import Transformer, WikiText2
from torch.utils.data import DataLoader
import time
import os

epochs = 1
epoch_size = 1000


class LanguageModel(L.LightningModule):
    def __init__(self, vocab_size):
        super().__init__()
        self.model = Transformer(vocab_size=vocab_size)
        self.t0 = 0
        self.epoch_time = []

    def on_train_start(self):
        # torch.set_num_threads(1)
        # torch.set_num_interop_threads(1)
        print(os.environ.get("OMP_NUM_THREADS"))
        print(torch.get_num_threads(), torch.get_num_interop_threads())

    def on_train_epoch_start(self):
        self.t0 = time.perf_counter()

    def on_train_epoch_end(self):
        self.epoch_time.append(time.perf_counter() - self.t0)

    def on_train_end(self):
        avg_epoch_time = sum(self.epoch_time) / len(self.epoch_time)
        it_per_sec = epochs * epoch_size / sum(self.epoch_time)
        self.trainer.print(f"avg time per epoch: {avg_epoch_time:.3f}")
        self.trainer.print(f"it/s: {it_per_sec:.3f}")

    def training_step(self, batch, batch_idx):
        input, target = batch
        output = self.model(input, target)
        loss = F.nll_loss(output, target.view(-1))
        self.log("train_loss", loss, prog_bar=True)
        return loss

    def configure_optimizers(self):
        return torch.optim.SGD(self.parameters(), lr=0.1)


def main():
    L.seed_everything(42)

    dataset = WikiText2()
    train_dataloader = DataLoader(dataset, batch_size=20, shuffle=True)
    model = LanguageModel(vocab_size=dataset.vocab_size)
    trainer = L.Trainer(gradient_clip_val=0.25, max_epochs=epochs, limit_train_batches=epoch_size, accelerator="cuda", devices=6, strategy="ddp")
    trainer.fit(model, train_dataloader)
    


if __name__ == "__main__":
    main()

```

DDP with 2 processes/GPUs

| num_threads | it/s |
| --- | --- |
| cpu_count | 32.481 |
| cpu_count / num_procs | 34.267 |
| 1 | 32.910 |

DDP with 6 processes/GPUs

| num_threads | it/s |
| --- | --- |
| cpu_count | 63.514 |
| cpu_count / num_procs | 65.260 |
| 1 | 64.409 |

## Discussion

In cases where CPU load is high like in the CV example where images need to be resized, the system can be overloaded if too many threads are launched. By restricting the upper bound on number of threads per GPU to num-cpus / num-procs, we ensure that the total number of threads across all processes in a machine is not exceedingly larger than the number of available cores on the system. This setting won't affect tasks that are CPU-light, as seen in the transformer example above. Torchrun/elastic sets the limit to 1 thread per DDP process, but warns that the value should be tuned. We could do the same, but in this PR we opt for a higher number and in return don't emit warnings. 

cc @carmocca @justusschock @awaelchli @borda

### CODE REVIEW

**Comment by Borda** [comment]
have you tried some extensive data preprocessing because this seems as you hitting IO bottleneck

File: src/lightning/fabric/utilities/distributed.py [hunk 1]
```diff
@@ -389,3 +390,16 @@ def __init__(self, sampler: Union[Sampler, Iterable], *args: Any, **kwargs: Any)
...
+    if num_processes < 1:
+        raise ValueError(f"`num_processes` should be >= 1, got {num_processes}.")
+    return max(1, _num_cpus_available() // num_processes)
+
+
+def _set_num_threads_if_needed(num_processes: int = 1) -> None:
+    if "OMP_NUM_THREADS" not in os.environ:
+        num_threads = _suggested_max_num_threads(num_processes)
+        torch.set_num_threads(num_threads)
+        os.environ["OMP_NUM_THREADS"] = str(num_threads)
```
Comment (carmocca) [inline src/lightning/fabric/utilities/distributed.py:405]: https://pytorch.org/docs/stable/notes/cpu_threading_torchscript_inference.html says that `set_num_threads` will ignore `OMP_NUM_THREADS` so maybe this line is not required.

That docs page talks about TorchScript but I think a lot of the information still applies.

**Review by carmocca — APPROVED** [review]

File: src/lightning/fabric/utilities/distributed.py (hunk 1 above)
Comment (awaelchli) [reply src/lightning/fabric/utilities/distributed.py:405]: It's not that it ignores it, but rather `set_num_threads` has precedence over the env var. But I'm setting it more for completeness rather than necessity (the user may look for this env var?). But I could also leave it out, I have no preference over which way here.

Comment (awaelchli) [reply src/lightning/fabric/utilities/distributed.py:405]: It's not that it ignores it, but rather `set_num_threads` has precedence over the env var. But I'm setting it more for completeness rather than necessity, so that the launched processes immediately get this setting. I could also leave it out, I don't mind.

**Comment by awaelchli** [comment]
@Borda I don't see how I'm hitting an IO bottleneck. The number of dataloading workers are the same between the comparison experiments, and so the IO load is identical. The only thing that changes between these runs is the number of threads used for the intra-op parallelism. The experiment simply shows that with choosing too many threads running in parallel, the effective throughput degrades.

**Comment by Borda** [comment]
> @Borda I don't see how I'm hitting an IO bottleneck. The number of dataloading workers are the same between the comparison experiments, and so the IO load is identical. The only thing that changes between these runs is the number of threads used for the intra-op parallelism. The experiment simply shows that with choosing too many threads running in parallel and competing with each other, the effective throughput degrades.

I do not say that your comparison is wrong or unfair, I just say that in case that when your preprocessing would take much longer for example image augmentation runs very often on CPU, then you can see some advantage of using small number of workers... the bottleneck IO I meant for example your IO speed os 10MB/s, then this speed is for all workers not each so it make sense when you divide 10MB/s per 5 workers they will be slover because each would have at most 2MB/s munis some switching overhead...

**Review by Borda — APPROVED** [review]
